File: app-rxjava/src/main/java/com/github/pwittchen/neurosky/app/rxjava/MainActivity.java
package com.github.pwittchen.neurosky.app.rxjava;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.util.Log;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import com.github.pwittchen.neurosky.library.RxNeuroSky;
import com.github.pwittchen.neurosky.library.message.enums.BrainWave;
import com.github.pwittchen.neurosky.library.message.enums.Signal;
import com.github.pwittchen.neurosky.library.message.enums.State;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import java.util.Locale;
import java.util.Set;

public class MainActivity extends AppCompatActivity {

  private final static String LOG_TAG = "NeuroSky";
  private RxNeuroSky neuroSky;
  private Disposable disposable;

  @BindView(R.id.tv_state) TextView tvState;
  @BindView(R.id.tv_attention) TextView tvAttention;
  @BindView(R.id.tv_meditation) TextView tvMeditation;
  @BindView(R.id.tv_blink) TextView tvBlink;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    ButterKnife.bind(this);
    neuroSky = new RxNeuroSky();
  }

  @Override protected void onResume() {
    super.onResume();
    disposable = neuroSky
        .stream()
        .subscribeOn(Schedulers.computation())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(brainEvent -> {
          handleStateChange(brainEvent.state());
          handleSignalChange(brainEvent.signal());
          handleBrainWavesChange(brainEvent.brainWaves());
        });
  }

  @Override protected void onPause() {
    super.onPause();
    if (disposable != null && !disposable.isDisposed()) {
      disposable.dispose();
    }
  }

  private void handleStateChange(final State state) {
    if (neuroSky != null && state.equals(State.CONNECTED)) {
      neuroSky.start();
    }

    tvState.setText(state.toString());
    Log.d(LOG_TAG, state.toString());
  }

  private void handleSignalChange(final Signal signal) {
    switch (signal) {
      case ATTENTION:
        tvAttention.setText(getFormattedMessage("attention: %d", signal));
        break;
      case MEDITATION:
        tvMeditation.setText(getFormattedMessage("meditation: %d", signal));
        break;
      case BLINK:
        tvBlink.setText(getFormattedMessage("blink: %d", signal));
        break;
    }

    Log.d(LOG_TAG, String.format("%s: %d", signal.toString(), signal.getValue()));
  }

  private String getFormattedMessage(String messageFormat, Signal signal) {
    return String.format(Locale.getDefault(), messageFormat, signal.getValue());
  }

  private void handleBrainWavesChange(final Set<BrainWave> brainWaves) {

    if (brainWaves.isEmpty()) {
      return;
    }

    StringBuilder stringBuilder = new StringBuilder();

    for (BrainWave bw : brainWaves) {
      String message = String.format(Locale.getDefault(), "%s: %d", bw.toString(), bw.getValue());
      stringBuilder.append(message).append(", ");
    }

    Log.d(LOG_TAG, stringBuilder.substring(0, stringBuilder.toString().length() - 2));
  }

  @SuppressLint("CheckResult") @OnClick(R.id.btn_connect) void connect() {
    neuroSky
        .connect()
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(
            () -> showMessage("connecting..."),
            throwable -> {
              showMessage(throwable.getMessage());
              Log.d(LOG_TAG, throwable.getMessage());
            });
  }

  @SuppressLint("CheckResult") @OnClick(R.id.btn_disconnect) void disconnect() {
    neuroSky
        .disconnect()
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(
            () -> showMessage("disconnected"),
            throwable -> showMessage(throwable.getMessage())
        );
  }

  @SuppressLint("CheckResult") @OnClick(R.id.btn_start_monitoring) void startMonitoring() {
    neuroSky
        .start()
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(
            () -> showMessage("started monitoring..."),
            throwable -> showMessage(throwable.getMessage())
        );
  }

  @SuppressLint("CheckResult") @OnClick(R.id.btn_stop_monitoring) void stopMonitoring() {
    neuroSky
        .stop()
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(
            () -> showMessage("stopped monitoring..."),
            throwable -> showMessage(throwable.getMessage())
        );
  }

  private void showMessage(String message) {
    Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
  }
}

File: app-rxkotlin/src/main/kotlin/com/github/pwittchen/neurosky/app/rxkotlin/MainActivity.kt
package com.github.pwittchen.neurosky.app.rxkotlin

import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.github.pwittchen.neurosky.library.RxNeuroSky
import com.github.pwittchen.neurosky.library.message.enums.BrainWave
import com.github.pwittchen.neurosky.library.message.enums.Signal
import com.github.pwittchen.neurosky.library.message.enums.State
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.btn_connect
import kotlinx.android.synthetic.main.activity_main.btn_disconnect
import kotlinx.android.synthetic.main.activity_main.btn_start_monitoring
import kotlinx.android.synthetic.main.activity_main.btn_stop_monitoring
import kotlinx.android.synthetic.main.activity_main.tv_attention
import kotlinx.android.synthetic.main.activity_main.tv_blink
import kotlinx.android.synthetic.main.activity_main.tv_meditation
import kotlinx.android.synthetic.main.activity_main.tv_state
import java.util.Locale

class MainActivity : AppCompatActivity() {

  companion object {
    const val LOG_TAG = "NeuroSky"
  }

  private lateinit var neuroSky: RxNeuroSky
  private lateinit var disposable: Disposable

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    neuroSky = RxNeuroSky()
    initButtonListeners()
  }

  private fun initButtonListeners() {
    btn_connect.setOnClickListener {
      neuroSky
          .connect()
          .subscribeOn(Schedulers.io())
          .observeOn(AndroidSchedulers.mainThread())
          .subscribe(
              { showMessage("connecting") },
              { throwable ->
                showMessage(throwable.message.toString())
                Log.d(LOG_TAG, throwable.message)
              })
    }

    btn_disconnect.setOnClickListener {
      neuroSky
          .disconnect()
          .subscribeOn(Schedulers.io())
          .observeOn(AndroidSchedulers.mainThread())
          .subscribe(
              { showMessage("disconnected") },
              { throwable -> showMessage(throwable.message.toString()) }
          )
    }

    btn_start_monitoring.setOnClickListener {
      neuroSky
          .start()
          .subscribeOn(Schedulers.io())
          .observeOn(AndroidSchedulers.mainThread())
          .subscribe(
              { showMessage("started monitoring...") },
              { throwable -> showMessage(throwable.message.toString()) }
          )
    }

    btn_stop_monitoring.setOnClickListener {
      neuroSky
          .stop()
          .subscribeOn(Schedulers.io())
          .observeOn(AndroidSchedulers.mainThread())
          .subscribe(
              { showMessage("stopped monitoring...") },
              { throwable -> showMessage(throwable.message.toString()) }
          )
    }
  }

  private fun showMessage(message: String) {
    Toast
        .makeText(this, message, Toast.LENGTH_SHORT)
        .show()
  }

  override fun onResume() {
    super.onResume()
    disposable = neuroSky
        .stream()
        .subscribeOn(Schedulers.computation())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe {
          handleStateChange(it.state())
          handleSignalChange(it.signal())
          handleBrainWavesChange(it.brainWaves())
        }
  }

  override fun onPause() {
    super.onPause()
    if (!disposable.isDisposed) {
      disposable.dispose()
    }
  }

  private fun handleStateChange(state: State) {
    if (state == State.CONNECTED) {
      neuroSky.start()
    }

    tv_state.text = state.toString()
    Log.d(LOG_TAG, state.toString())
  }

  private fun handleSignalChange(signal: Signal) {
    when (signal) {
      Signal.ATTENTION -> tv_attention.text = getFormattedMessage("attention: %d", signal)
      Signal.MEDITATION -> tv_meditation.text = getFormattedMessage("meditation: %d", signal)
      Signal.BLINK -> tv_blink.text = getFormattedMessage("blink: %d", signal)
      else -> Log.d(LOG_TAG, "unhandled signal")
    }

    Log.d(LOG_TAG, String.format("%s: %d", signal.toString(), signal.value))
  }

  private fun getFormattedMessage(
    messageFormat: String,
    signal: Signal
  ): String {
    return String.format(Locale.getDefault(), messageFormat, signal.value)
  }

  private fun handleBrainWavesChange(brainWaves: Set<BrainWave>) {

    if (brainWaves.isEmpty()) {
      return
    }

    val stringBuilder = StringBuilder()

    for (bw in brainWaves) {
      val message = String.format(Locale.getDefault(), "%s: %d", bw.toString(), bw.value)
      stringBuilder
          .append(message)
          .append(", ")
    }

    Log.d(LOG_TAG, stringBuilder.substring(0, stringBuilder.toString().length - 2))
  }
}
